src/compiler/types.rs

```diff
@@ -1157,60 +1202,455 @@ bitflags! {
         self.flags.get()
     }
 
-    pub fn set_flags(&self, flags: SymbolFlags) {
+    fn set_flags(&self, flags: SymbolFlags) {
         self.flags.set(flags);
     }
 
-    pub fn maybe_declarations(&self) -> Ref<Option<Vec<Rc<Node>>>> {
+    fn escaped_name(&self) -> &__String {
+        &self.escaped_name
+    }
+
+    fn maybe_declarations(&self) -> Ref<Option<Vec<Rc<Node>>>> {
         self.declarations.borrow()
     }
 
-    pub fn set_declarations(&self, declarations: Vec<Rc<Node>>) {
+    fn set_declarations(&self, declarations: Vec<Rc<Node>>) {
         *self.declarations.borrow_mut() = Some(declarations);
     }
 
-    pub fn maybe_value_declaration(&self) -> Ref<Option<Weak<Node>>> {
+    fn maybe_value_declaration(&self) -> Ref<Option<Weak<Node>>> {
         self.value_declaration.borrow()
     }
 
-    pub fn set_value_declaration<TNode: NodeInterface>(&self, node: &TNode) {
-        *self.value_declaration.borrow_mut() = Some(Rc::downgrade(&node.node_wrapper()));
+    fn set_value_declaration(&self, node: Rc<Node>) {
+        *self.value_declaration.borrow_mut() = Some(Rc::downgrade(&node));
     }
 
-    pub fn maybe_members(&self) -> RefMut<Option<Rc<RefCell<SymbolTable>>>> {
+    fn maybe_members(&self) -> RefMut<Option<Rc<RefCell<SymbolTable>>>> {
         self.members.borrow_mut()
     }
 
-    pub fn members(&self) -> Rc<RefCell<SymbolTable>> {
+    fn members(&self) -> Rc<RefCell<SymbolTable>> {
         self.members.borrow_mut().as_ref().unwrap().clone()
     }
+
+    fn maybe_id(&self) -> Option<SymbolId> {
+        self.id.get()
+    }
+
+    fn id(&self) -> SymbolId {
+        self.id.get().unwrap()
+    }
+
+    fn set_id(&self, id: SymbolId) {
+        self.id.set(Some(id));
+    }
+}
+
+impl From<BaseSymbol> for Symbol {
+    fn from(base_symbol: BaseSymbol) -> Self {
+        Symbol::BaseSymbol(base_symbol)
+    }
+}
+
+#[derive(Debug)]
+pub struct SymbolLinks {
+    pub target: Option<Rc<Symbol>>,
+    pub type_: Option<Rc<Type>>,
+    pub declared_type: Option<Rc<Type>>,
+    pub mapper: Option<TypeMapper>,
+}
+
+impl SymbolLinks {
+    pub fn new() -> Self {
+        Self {
+            target: None,
+            type_: None,
+            declared_type: None,
+            mapper: None,
+        }
+    }
+}
+
+bitflags! {
+    pub struct CheckFlags: u32 {
+        const None = 0;
+        const Instantiated = 1 << 0;
+        const SyntheticProperty = 1 << 1;
+        const SyntheticMethod = 1 << 2;
+        const Readonly = 1 << 3;
+        const Late = 1 << 12;
+        const OptionalParameter = 1 << 14;
+        const RestParameter = 1 << 15;
+
+        const Synthetic = Self::SyntheticProperty.bits | Self::SyntheticMethod.bits;
+    }
+}
+
+pub trait TransientSymbolInterface: SymbolInterface {
+    fn symbol_links(&self) -> Rc<RefCell<SymbolLinks>>;
+    fn check_flags(&self) -> CheckFlags;
+}
+
+#[derive(Debug)]
+pub enum TransientSymbol {
+    BaseTransientSymbol(BaseTransientSymbol),
```

Comment:
It looks like there will be other sub-variants of `TransientSymbol` so made it an enum